File: src/java/ironfist/persistence/Persistence.java
package ironfist.persistence;

import ironfist.persistence.file.FileLog;
import ironfist.persistence.file.FileStore;
import ironfist.util.Closure;

import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.util.HashMap;

public final class Persistence {

  private static File dataFile;
  private static File logFile;
  private static Store store;
  private static Log log;
  private static Engine persistence;

  private Persistence() {

  }

  public static void create(String name) throws PersistenceException {
    File directory = new File(System.getProperty("user.home"), "ironfist");
    directory.mkdirs();

    if (directory.listFiles(new Filter(name)).length > 0) {
      throw new PersistenceException("repository " + name + " already exists");
    }

    dataFile = new File(directory, name + ".dat");
    logFile = new File(directory, name + ".log");
    try {
      store = new FileStore(dataFile);
      log = new FileLog(logFile);
      persistence = new Engine(store, log);
      persistence.update(new Create());
      persistence.checkpoint();
    } catch (Exception e) {
      new PersistenceException(e.getMessage());
    }
  }

  public static void open(String name) throws PersistenceException {
    File directory = new File(System.getProperty("user.home"), "ironfist");

    if (directory.listFiles(new Filter(name)).length > 0) {
      throw new PersistenceException("repository " + name + " does not exist");
    }

    dataFile = new File(directory, name + ".dat");
    logFile = new File(directory, name + ".log");
    try {
      store = new FileStore(dataFile);
      log = new FileLog(logFile);
      persistence = new Engine(new FileStore(dataFile), new FileLog(logFile));
    } catch (Exception e) {
      new PersistenceException(e.getMessage());
    }
  }

  public static void put(Object key, Object value) throws PersistenceException {
    try {
      persistence.update(new Put(key, value));
      persistence.checkpoint();
    } catch (IOException e) {
      throw new PersistenceException("error updating root");
    }
  }

  public static Object get(Object key) throws PersistenceException {
    Object result = null;

    try {
      result = persistence.query(new Get(key));
    } catch (Exception e) {
      throw new PersistenceException("error loading object");
    }

    return result;
  }

  public static void close() {
  }

  public static void delete(String name) {
    try {
      persistence.close();
    } catch (IOException e) {
      e.printStackTrace();
    }

    File directory = new File(System.getProperty("user.home"), "ironfist");
    File[] files = directory.listFiles(new Filter(name));

    for (int index = 0; index < files.length; index++) {
      files[index].delete();
    }
  }

  public static String[] list() {
    return null;
  }

  private static class Filter implements FilenameFilter {

    private String name;

    public Filter(String name) {
      this.name = name;
    }

    public boolean accept(File dir, String target) {
      return target.startsWith(name + ".");
    }
  }

  private static class Create implements Closure<Reference, Object> {

    private static final long serialVersionUID = 3690476935130198069L;

    public Object apply(Reference reference) {
      reference.set(new HashMap<Object, Object>());
      return null;
    }
  }

  private static class Get implements Closure<Reference, Object> {

    private static final long serialVersionUID = 3258411724993474867L;
    private Object key;

    public Get(Object key) {
      this.key = key;
    }

    public Object apply(Reference reference) {
      return ((HashMap<Object, Object>) reference.get()).get(key);
    }
  }

  private static class Put implements Closure<Reference, Object> {

    private static final long serialVersionUID = 3256727294637979448L;
    private Object key;
    private Object value;

    public Put(Object key, Object value) {
      this.key = key;
      this.value = value;
    }

    public Object apply(Reference reference) {
      return ((HashMap<Object, Object>) reference.get()).put(key, value);
    }
  }

}

File: src/main/java/ukalus/persistence/Persistence.kt
package ukalus.persistence

import ukalus.persistence.file.FileLog
import ukalus.persistence.file.FileStore
import java.io.File
import java.io.FilenameFilter
import java.io.Serializable
import java.util.*
import java.util.function.Function

object Persistence {

    lateinit private var dataFile: File
    lateinit private var logFile: File
    lateinit private var store: Store<HashMap<Any, Any>>
    lateinit private var log: Log<Function<Reference<HashMap<Any, Any>?>, Any?>>
    private var persistence: Engine<HashMap<Any, Any>>? = null

    @Throws(PersistenceException::class)
    fun create(name: String) {
        val directory = File(System.getProperty("user.home"), "ukalus")
        directory.mkdirs()

        val files = directory.listFiles(Filter(name))
        if (files.isNotEmpty()) {
            throw PersistenceException(String.format("repository %s already exists", name))
        }

        dataFile = File(directory, String.format("%s.dat", name))
        logFile = File(directory, String.format("%s.log", name))
        store = FileStore<HashMap<Any, Any>>(dataFile)
        log = FileLog<Function<Reference<HashMap<Any, Any>?>, Any?>>(logFile)
        persistence = Engine(store, log)
        persistence!!.update(Create())
        persistence!!.checkpoint()

    }

    @Throws(PersistenceException::class)
    fun open(name: String) {
        val directory = File(System.getProperty("user.home"), "ukalus")

        val files = directory.listFiles(Filter(name))
        if (files.isNotEmpty()) {
            throw PersistenceException(String.format("repository %s does not exist", name))
        }

        dataFile = File(directory, String.format("%s.dat", name))
        logFile = File(directory, String.format("%s.log", name))
        store = FileStore<HashMap<Any, Any>>(dataFile)
        log = FileLog<Function<Reference<HashMap<Any, Any>?>, Any?>>(logFile)
        persistence = Engine(store, log)

    }

    @Throws(PersistenceException::class)
    fun put(key: Any, value: Any) {
        persistence!!.update(Put(key, value))
        persistence!!.checkpoint()
    }

    @Throws(PersistenceException::class)
    operator fun get(key: Any): Any? {
        return persistence!!.query(Get(key))
    }

    fun close() {}

    fun delete(name: String) {
        persistence!!.close()

        val directory = File(System.getProperty("user.home"), "ukalus")
        val files = directory.listFiles(Filter(name))

        for (file in files!!) {
            file.delete()
        }
    }

    fun list(): Array<String>? {
        return null
    }

    private class Filter(private val name: String) : FilenameFilter {

        override fun accept(dir: File, target: String): Boolean {
            return target.startsWith(name + ".")
        }
    }

    private class Create : Function<Reference<HashMap<Any, Any>?>, Any?>, Serializable {

        override fun apply(reference: Reference<HashMap<Any, Any>?>): Any? {
            reference.set(HashMap<Any, Any>())
            return null
        }
    }

    private class Get(private val key: Any) : Function<Reference<HashMap<Any, Any>?>, Any?>, Serializable {

        override fun apply(reference: Reference<HashMap<Any, Any>?>): Any? {
            return reference.get()?.get(key)
        }
    }

    private class Put(private val key: Any, private val value: Any) : Function<Reference<HashMap<Any, Any>?>, Any?>, Serializable {

        override fun apply(reference: Reference<HashMap<Any, Any>?>): Any? {
            return reference.get()?.put(key, value)
        }
    }

}